File: src/odgs/executive/interceptor.py
import json
import os
import re
import sys
import logging
import datetime
from typing import Dict, List, Any

# Add project root to sys.path to allow imports from odgs.system
current_dir = os.path.dirname(os.path.abspath(__file__))
project_root = os.path.dirname(current_dir)
if project_root not in sys.path:
    sys.path.append(project_root)

from odgs.system.scripts.hashing import generate_project_hash

# --- ARTICLE 12: IMMUTABLE AUDIT LOGGING ---
# In a real Sovereign Cloud, this would write to a WORM (Write Once Read Many) drive.
# Here we simulate it with a standard file logger.
audit_logger = logging.getLogger("sovereign_audit")
audit_logger.setLevel(logging.INFO)
# Ensure we don't add multiple handlers on reload
if not audit_logger.handlers:
    handler = logging.FileHandler(os.path.join(project_root, "sovereign_audit.log"))
    handler.setFormatter(logging.Formatter('%(asctime)s - %(message)s'))
    audit_logger.addHandler(handler)
# -------------------------------------------

class ProcessBlockedException(Exception):
    """Raised when a Process is blocked by a Rule Violation (Hard Stop)."""
    pass

class SecurityException(Exception):
    """Raised when a Cryptographic Handshake failure occurs."""
    pass

class OdgsInterceptor:
    def __init__(self, project_root_path: str = None):
        """
        Initialize the Interceptor.
        :param project_root_path: Path to the project root. 
                                  If None, attempts to find it relative to this file.
        """
        if project_root_path:
            self.project_root = project_root_path
        else:
            # We are in /executive/interceptor.py, so root is one level up
            self.project_root = os.path.dirname(os.path.abspath(__file__))
            if self.project_root.endswith("executive"):
                self.project_root = os.path.dirname(self.project_root)

        self.graph = self._load_from_plane("legislative", "ontology_graph.json")
        self.rules = self._load_rules()
        self.metrics = self._load_from_plane("legislative", "standard_metrics.json")
    
    def _load_from_plane(self, plane: str, filename: str) -> Dict:
        """
        Load JSON artifacts from their specific Sovereign Plane.
        """
        path = os.path.join(self.project_root, plane, filename)
        if not os.path.exists(path):
            raise FileNotFoundError(f"Sovereign Artifact not found in {plane} plane: {path}")
        with open(path, 'r') as f:
            return json.load(f)

    def _load_rules(self) -> Dict[str, Dict]:
        """Load rules from Judiciary Plane and index them by URN."""
        rules_data = self._load_from_plane("judiciary", "standard_data_rules.json")
        
        if isinstance(rules_data, dict) and "rules" in rules_data:
            rules_list = rules_data["rules"]
        elif isinstance(rules_data, list):
            rules_list = rules_data
        elif isinstance(rules_data, dict):
            rules_list = [rules_data]
        else:
            rules_list = []

        indexed = {}
        for rule in rules_list:
             rid = str(rule.get("rule_id", ""))
             urn = f"urn:odgs:rule:{rid}"
             indexed[urn] = rule
        return indexed

    def intercept(self, process_urn: str, data_context: Dict[str, Any], required_integrity_hash: str) -> bool:
        """
        The Core Logic: Checks if the requested Process Stage is blocked by any Rules.
        
        STRICT MODE ENFORCED:
        - required_integrity_hash is MANDATORY.
        - All decisions are logged to sovereign_audit.log.
        """
        
        # 1. ARTICLE 12 LOGGING (Start)
        audit_entry = {
            "timestamp": datetime.datetime.utcnow().isoformat() + "Z",
            "process_urn": process_urn,
            "input_hash": required_integrity_hash,
            "status": "PENDING"
        }

        print(f"🛡️  Interceptor Active: Checking Access for Process [{process_urn}]")

        try:
            # 2. SOVEREIGN HANDSHAKE (Cryptographic Verification)
            # This is now MANDATORY. The "Doorman" checks ID every time.
            print(f"   🔐 Handshake Requested. Verifying Integrity...")
            current_state = generate_project_hash(self.project_root)
            current_hash = current_state["master_hash"]
            
            if current_hash != required_integrity_hash:
                audit_entry["status"] = "REJECTED_HASH_MISMATCH"
                audit_entry["actual_hash"] = current_hash
                audit_logger.error(json.dumps(audit_entry))
                
                error_msg = (
                    f"CRITICAL SECURITY FAILURE: Governance Hash Mismatch.\n"
                    f"Expected: {required_integrity_hash}\n"
                    f"Actual:   {current_hash}\n"
                    f"The Semantic Laws have been altered or do not match the signature."
                )
                print(f"   ⛔ {error_msg}")
                raise SecurityException(error_msg)
            else:
                print(f"   ✅ Integrity Verified. Hash matches.")

            # 3. IDENTIFY BLOCKING RULES
            blocking_rules = self._find_blocking_rules(process_urn)
            
            if not blocking_rules:
                print("   ✅ No Hard Stop rules found for this process.")
                audit_entry["status"] = "APPROVED_NO_RULES"
                audit_logger.info(json.dumps(audit_entry))
                return True

            print(f"   ⚠️  Found {len(blocking_rules)} Potential Blocking Rules.")

            # 4. EVALUATE BLOCKING RULES
            for rule_urn in blocking_rules:
                rule_def = self.rules.get(rule_urn)
                if not rule_def:
                    print(f"   [warn] Rule definition not found for {rule_urn}, skipping.")
                    continue
                
                self._evaluate_rule(rule_def, data_context)
            
            print("   ✅ All Blocking Rules Passed. Access Granted.")
            audit_entry["status"] = "APPROVED"
            audit_logger.info(json.dumps(audit_entry))
            return True

        except Exception as e:
            # Catch Validation Errors and Log them as Rejections
            if audit_entry["status"] == "PENDING":
                 audit_entry["status"] = "REJECTED_RULE_VIOLATION"
            
            audit_entry["error_type"] = type(e).__name__
            audit_entry["error_msg"] = str(e)
            audit_logger.error(json.dumps(audit_entry))
            raise e

    def _find_blocking_rules(self, target_process_urn: str) -> List[str]:
        blockers = []
        edges = self.graph.get("graph_edges", [])
        for edge in edges:
            if (edge.get("target_urn") == target_process_urn and 
                edge.get("relationship") == "BLOCKS_PROCESS"):
                blockers.append(edge.get("source_urn"))
        return blockers

    def _evaluate_rule(self, rule_def: Dict, data: Dict):
        rule_id = rule_def.get("rule_id")
        rule_name = rule_def.get("name")
        print(f"      🔎 Verifying Rule {rule_id}: {rule_name}")

        # Logic for Rule 2021: ISO 6346 Container ID
        if str(rule_id) == "2021":
            cid = data.get("container_id", "")
            pattern = r"^[A-Z]{4}[0-9]{7}$"
            if not re.match(pattern, str(cid)):
                raise ProcessBlockedException(
                    f"HARD STOP: Rule {rule_id} Failed. "
                    f"Value '{cid}' does not match pattern {pattern}. "
                    f"Process is BLOCKED."
                )
        else:
             print(f"      [info] No executable logic implemented for Rule {rule_id}, passing by default.")

File: lib/interceptor.js
import ontologyGraph from './schemas/legislative/ontology_graph.json' with { type: "json" };
import standardDataRules from './schemas/judiciary/standard_data_rules.json' with { type: "json" };
import { Parser } from 'expr-eval';

// ─── Sovereign Handshake (Node.js) ──────────────────────────────
// Computes SHA-256 hashes of legislative JSON files for integrity verification.
// Mirrors the Python `generate_project_hash()` implementation.

let _crypto = null;
let _fs = null;
let _path = null;

try {
    _crypto = await import('node:crypto');
    _fs = await import('node:fs');
    _path = await import('node:path');
} catch (e) {
    // Browser environment — crypto/fs/path not available, handshake disabled.
}

function computeDefinitionHash(schemasDir) {
    if (!_crypto || !_fs || !_path) return null;

    const schemaFiles = [
        'legislative/standard_metrics.json',
        'legislative/standard_dq_dimensions.json',
        'legislative/ontology_graph.json',
        'judiciary/standard_data_rules.json',
        'judiciary/root_cause_factors.json',
        'executive/business_process_maps.json',
        'executive/physical_data_map.json'
    ].sort();

    let comboString = '';

    for (const relPath of schemaFiles) {
        const fullPath = _path.join(schemasDir, relPath);
        try {
            const data = JSON.parse(_fs.readFileSync(fullPath, 'utf-8'));
            const canonical = JSON.stringify(data, Object.keys(data).sort(), 0)
                .replace(/\s+/g, ''); // Compact canonical form
            // Re-parse for deterministic key sorting
            const sorted = JSON.stringify(JSON.parse(canonical), null, 0);
            const hash = _crypto.createHash('sha256').update(sorted).digest('hex');
            comboString += hash;
        } catch (e) {
            console.warn(`Schema file load failed: ${relPath}: ${e.message}`);
            comboString += 'MISSING_FILE';
        }
    }

    return _crypto.createHash('sha256').update(comboString).digest('hex');
}

// ─── Exception Classes ──────────────────────────────────────────

export class ProcessBlockedException extends Error {
    constructor(message) {
        super(message);
        this.name = "ProcessBlockedException";
    }
}

export class SecurityException extends Error {
    constructor(message) {
        super(message);
        this.name = "SecurityException";
    }
}

// ─── Safe Expression Evaluator ──────────────────────────────────

const parser = new Parser();

// Register custom functions for rule evaluation
function regex_match(pattern, val) {
    if (!val) return false;
    try {
        return new RegExp(pattern).test(String(val));
    } catch (e) {
        console.warn(`regex_match failed: ${e.message}`);
        return false;
    }
}

function parse_date(val) {
    if (!val) return new Date(-8640000000000000).getTime();
    return new Date(val).getTime();
}

function today_fn() {
    const d = new Date();
    d.setHours(0, 0, 0, 0);
    return d.getTime();
}

// ─── Main Interceptor ───────────────────────────────────────────

export class OdgsInterceptor {
    constructor(schemasDir = null) {
        this.graph = ontologyGraph;
        this.rules = this._loadRules();
        this.schemasDir = schemasDir;
    }

    _loadRules() {
        const indexed = {};
        let rulesList = [];

        if (Array.isArray(standardDataRules)) {
            rulesList = standardDataRules;
        } else if (standardDataRules.rules && Array.isArray(standardDataRules.rules)) {
            rulesList = standardDataRules.rules;
        } else {
            rulesList = [standardDataRules];
        }

        rulesList.forEach(rule => {
            const rid = rule.rule_id;
            const urn = rule.urn || `urn:odgs:rule:${rid}`;
            indexed[urn] = rule;
        });
        return indexed;
    }

    /**
     * Intercepts data flow and enforces sovereign binding (v3.3).
     * @param {string} processUrn - The process lifecycle stage (e.g., "urn:odgs:process:O2C_S03").
     * @param {object} dataContext - The payload to validate.
     * @param {string|null} requiredIntegrityHash - The expected SHA-256 hash of the Legislative definition.
     * @returns {boolean} true if all rules pass.
     * @throws {ProcessBlockedException} if any HARD_STOP rule fails.
     * @throws {SecurityException} if the Sovereign Handshake fails.
     */
    intercept(processUrn, dataContext, requiredIntegrityHash = null) {

        // 1. SOVEREIGN HANDSHAKE — Validate Legislative Integrity
        if (requiredIntegrityHash && this.schemasDir) {
            const actualHash = computeDefinitionHash(this.schemasDir);
            if (actualHash && requiredIntegrityHash !== actualHash) {
                throw new SecurityException(
                    `Sovereign Handshake Failed: Legislative artifacts have been modified. ` +
                    `Expected hash ${requiredIntegrityHash.slice(0, 16)}..., ` +
                    `got ${actualHash.slice(0, 16)}...`
                );
            }
        }

        // 2. Identify Blocking Rules from the Graph
        const blockingRules = this._findBlockingRules(processUrn);

        if (blockingRules.length === 0) {
            return true;
        }

        // 3. Evaluate each Blocking Rule
        const violations = [];
        const warnings = [];

        for (const ruleUrn of blockingRules) {
            const ruleDef = this.rules[ruleUrn];
            if (!ruleDef) continue;

            const result = this._evaluateRule(ruleDef, dataContext);
            if (result.violation) {
                const severity = ruleDef.severity || "HARD_STOP";
                if (severity === "HARD_STOP") {
                    violations.push(result.violation);
                } else if (severity === "WARNING") {
                    warnings.push(result.violation);
                }
            }
        }

        // 4. ENFORCE
        if (violations.length > 0) {
            throw new ProcessBlockedException(
                `Governance Failure: ${JSON.stringify(violations)}`
            );
        }

        return true;
    }

    _findBlockingRules(targetProcessUrn) {
        const edges = this.graph.graph_edges || [];
        return edges
            .filter(edge => edge.target_urn === targetProcessUrn && edge.relationship === "BLOCKS_PROCESS")
            .map(edge => edge.source_urn);
    }

    /**
     * Evaluates a single rule against data using expr-eval (safe, no eval/Function).
     * @returns {{ violation: string|null }}
     */
    _evaluateRule(ruleDef, data) {
        const ruleId = ruleDef.rule_id;
        const ruleName = ruleDef.name;
        const logic = ruleDef.logic_expression;

        if (logic) {
            try {
                // Transpile Python boolean operators to JS-compatible
                let jsLogic = logic
                    .replace(/\band\b/g, "and")
                    .replace(/\bor\b/g, "or")
                    .replace(/\bnot\b/g, "not")
                    .replace(/r'([^']*)'/g, "'$1'")
                    .replace(/r"([^"]*)"/g, '"$1"');

                // Build evaluation context
                const evalContext = {
                    ...data,
                    value: data.value !== undefined
                        ? data.value
                        : (Object.keys(data).length === 1 ? Object.values(data)[0] : undefined),
                };

                // Register custom functions on the parser
                const p = new Parser();
                p.functions.regex_match = regex_match;
                p.functions.parse_date = parse_date;
                p.functions.today = today_fn;

                const isValid = p.evaluate(jsLogic, evalContext);

                if (!isValid) {
                    return { violation: `Rule ${ruleId} Failed: ${ruleName}` };
                }

            } catch (e) {
                // "Fail Closed" — treat execution errors as violations
                return { violation: `Rule ${ruleId} Execution Error: ${e.message}` };
            }
            return { violation: null };
        }

        // Legacy hardcoded fallback for Rule 2021 (Container ID Format)
        if (String(ruleId) === "2021") {
            const cid = data.container_id || "";
            const pattern = /^[A-Z]{4}[0-9]{7}$/;
            if (!pattern.test(cid)) {
                return { violation: `Rule ${ruleId} Failed: Value '${cid}' does not match pattern.` };
            }
        }

        return { violation: null };
    }
}
